### PR TITLE
21 batch update actions

### PR DESCRIPTION
**Pull request by heshamMassoud**
#### Summary
This PR addresses mainly #21. Since CTP only allows 500 update actions as a maximum number of actions per update request. Therefore, the sync needs to handle requests which could have more than 500 by batching update requests in batches of 500 update actions.

#### Detailed Hints for review
1. Change of method name [`batchDrafts` to `batchElements`](https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-f36b6d323158c6e3371f2f2d80ce0ea9) which was used for batching drafts only, but now can be used in anything (incl. batching update actions)
2. Method usage name changes: 
     - https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-44d2fa184e625bcc0900b105f60d8677
     - https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-b70cdd276f9a92a4eabc591e1f9f9011L223
     - https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-dfaa8dde01a9924251be828b4bdd0a1e

5. update action batching
     - Implementation in [BaseService](https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-b89f1861d73d97f8e3869436b707f662).
     - Usage of `updateResource` method in services: [CategoryServiceImpl](https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-38b71d2c035833260351074420bf8067), and [ProductServiceImpl](https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-6dfa47f1d198543a343a810f8a0c0ab5)
     - IT test for update action batching: [`updateProduct_WithMoreThan500Actions_ShouldNotFail `](https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-4fcf28164f083f7b0c50764b1a938399R589)
     - Since now the update service returns a CompletionStage resulting from a [.thenCompose](https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-b89f1861d73d97f8e3869436b707f662R104) always because of the batching. Therefore, if a sphere exception is thrown it will be wrapped in a `CompletionException`. Therefore, checking for concurrent modification exception resulting from update changes to checking the cause of the exception: https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-365abd2d37edf8522541a0b8b126963f and unit tests were adapted: https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-45ac84a99656265869fd7f1e61d3ecdc

2. Boy-scout errands:
     - Change of assertions from error messages to stat counters: 
          - https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-b70cdd276f9a92a4eabc591e1f9f9011L117
          -  https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-d68ebb35acffe45efc007ae884c446b0
     - This IT [`updateProduct_WithInvalidChanges_ShouldNotUpdateProduct`](https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-4fcf28164f083f7b0c50764b1a938399L669) was just moved up to a correct position in the test class. So no impl changes in it. 

#### Relevant Issues
#21 

#### Todo
- Tests
    - [x] Unit 
    - [x] Integration
- [x] Documentation
- [x] Add Release Notes entry.: 6aaaa9b

### CODE REVIEW

File: src/main/java/com/commercetools/sync/commons/utils/SyncUtils.java [hunk 1]
```diff
@@ -15,21 +15,19 @@
...
      */
-    public static <T> List<List<T>> batchDrafts(@Nonnull final List<T> drafts,
-                                                final int batchSize) {
+    public static <T> List<List<T>> batchElements(@Nonnull final List<T> elements, final int batchSize) {
         List<List<T>> batches = new ArrayList<>();
-        for (int i = 0; i < drafts.size() && batchSize > 0; i += batchSize) {
-            batches.add(drafts.subList(i, Math.min(i + batchSize, drafts.size())));
+        for (int i = 0; i < elements.size() && batchSize > 0; i += batchSize) {
+            batches.add(elements.subList(i, Math.min(i + batchSize, elements.size())));
         }
```
Comment (butenkor) [inline src/main/java/com/commercetools/sync/commons/utils/SyncUtils.java]: why not `Lists.partition(elements, batchSize)`?

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java [hunk 2]
```diff
@@ -28,6 +32,7 @@
     final Map<String, String> keyToIdCache = new ConcurrentHashMap<>();
 
     private static final String CREATE_FAILED = "Failed to create draft with key: '%s'. Reason: %s";
+    private static final int MAXIMUM_ALLOWED_UPDATE_ACTIONS = 500;
```
Comment (butenkor) [inline src/main/java/com/commercetools/sync/services/impl/BaseService.java]: Default is good but would be nice if one could override it in syncOptions.

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java [hunk 3]
```diff
@@ -78,4 +83,55 @@
...
+        @Nonnull final List<List<UpdateAction<U>>> batches) {
+        if (batches.isEmpty()) {
+            return result;
+        }
+        final List<UpdateAction<U>> nextBatch = batches.remove(0);
+        final CompletionStage<U> nextBatchUpdateFuture = result.thenCompose(updatedProduct ->
+            syncOptions.getCtpClient().execute(updateCommandFunction.apply(updatedProduct, nextBatch))
+        );
+        return updateBatches(nextBatchUpdateFuture, updateCommandFunction, batches);
+    }
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/sync/services/impl/BaseService.java]: i see here couple related problems:
1. Why did you decide to use recursion? Does it give you any benefit here?
2. `remove(0)` is quite inefficient operation on random access lists, like `ArrayList`: it rebuilds the whole least every time. Of course, for short lists (like in this case) it is not a bid deal, but anyway try not to use such inconvenient approaches in our code.
3. We have a result, which we wrap to `CompletionStage` and then immediately compose -> unnecessary wrapping. Wrap it to `CompletionStage` only when return the result in `if (batches.isEmpty())`

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java [hunk 4]
```diff
@@ -78,4 +83,55 @@
...
+     */
+    @Nonnull
+    private CompletionStage<U> updateBatches(
+        @Nonnull final CompletionStage<U> result,
+        @Nonnull final BiFunction<U, List<? extends UpdateAction<U>>, UpdateCommand<U>> updateCommandFunction,
+        @Nonnull final List<List<UpdateAction<U>>> batches) {
+        if (batches.isEmpty()) {
+            return result;
+        }
+        final List<UpdateAction<U>> nextBatch = batches.remove(0);
```
Comment (butenkor) [inline src/main/java/com/commercetools/sync/services/impl/BaseService.java]: Should it be removed rather after successful execute?

File: src/main/java/com/commercetools/sync/commons/utils/SyncUtils.java (hunk 1 above)
Comment (andrii-kovalenko-ct) [reply src/main/java/com/commercetools/sync/commons/utils/SyncUtils.java]: guess because Guava is not included to the project

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java (hunk 3 above)
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/services/impl/BaseService.java]: 1. I have used recursion not because it gives me any benefit other than it was the most straightforward and cleanest way for me to implement processing of on an array of completion stages that depend the previous element's completion stage :) better suggestions?
2. IMO, it is completely justified in this case with the worst case complexity being O(n) for removal (with the shifting) on the most probably very small sizes of batches. (the update actions are only per one product (comparing one product to another). Also I am not sure, how to be better. I mean it could be O(1) If I remove the last batch first, but then to do that I would need to reverse the list (and that is also linear O(n) ).
3. I really don't see how it's unnecessary. (Or maybe I don't understand this point at all :) ) But I the completion stage is there because its needed for the recursive call so that the subsequent calls are only executed after the execution of the earlier completion stages.

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java (hunk 4 above)
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/services/impl/BaseService.java]: If it failed to execute the batch, It will not continue with the rest of update action batches anyway. So it doesn't make a difference if was removed then succeeded or failed. The list of batches will be obsolete after failure.

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java (hunk 2 above)
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/services/impl/BaseService.java]: You mean have it as one of the options in the sync? What are the use cases?

File: src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/categories/CategorySyncIT.java [hunk 5]
```diff
@@ -114,10 +114,10 @@ public void syncDrafts_withChangesOnly_ShouldUpdateCategories() {
...
         final CategorySyncStatistics syncStatistics = categorySync.sync(categoryDrafts).toCompletableFuture().join();
 
-        assertThat(syncStatistics.getReportMessage())
-            .isEqualTo(format("Summary: %d categories were processed in total (%d created, %d updated, %d failed to "
-                + "sync and %s categories with a missing parent).", 2, 0, 2, 0, 0));
-
+        assertThat(syncStatistics.getProcessed()).isEqualTo(2);
+        assertThat(syncStatistics.getCreated()).isEqualTo(0);
+        assertThat(syncStatistics.getUpdated()).isEqualTo(2);
+        assertThat(syncStatistics.getFailed()).isEqualTo(0);
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/categories/CategorySyncIT.java]: There is an issue for this specifically actually: https://github.com/commercetools/commercetools-sync-java/issues/186

But I just change it whenever I find it on the way for now ;)

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java (hunk 3 above)
Comment (andrii-kovalenko-ct) [reply src/main/java/com/commercetools/sync/services/impl/BaseService.java]: 1. IMHO recursion is always confusing, especially if the data structure is not recursive (like graphs), and usually more error prone.
2. Question is not about complexity, but about code style - removing zero element looks suspicious and usually used in some complex logic
3. Then it is one more point to avoid recursion :)

File: src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java [hunk 6]
```diff
@@ -545,6 +548,77 @@ public void updateProduct_WithValidChanges_ShouldUpdateProductCorrectly() {
...
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+        assertThat(fetchedProductOptional).isNotEmpty();
+        final Product fetchedProduct = fetchedProductOptional.get();
+        assertThat(fetchedProduct.getMasterData().getCurrent().getName())
+            .isEqualTo(updatedProduct.getMasterData().getCurrent().getName());
+        assertThat(fetchedProduct.getMasterData().getCurrent().getSlug())
+            .isEqualTo(updatedProduct.getMasterData().getCurrent().getSlug());
+        assertThat(fetchedProduct.getKey()).isEqualTo(updatedProduct.getKey());
+    }
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: I made the assertions more specific here thx: e6c2f94

About the edge cases you mentioned, why are they edge cases? and how are they not covered by that test?

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java (hunk 3 above)
Comment (andrii-kovalenko-ct) [reply src/main/java/com/commercetools/sync/services/impl/BaseService.java]: My proposition:
```java 
private CompletionStage<U> updateBatches(
    @Nonnull final CompletionStage<U> result,
    @Nonnull final BiFunction<U, List<? extends UpdateAction<U>>, UpdateCommand<U>> updateCommandFunction,
    @Nonnull final List<List<UpdateAction<U>>> batches) {
    if (batches.isEmpty()) {
        return result;
    }

    CompletionStage<U> completionStage = result;

    for (final List<UpdateAction<U>> batch : batches) {
        completionStage = completionStage.thenCompose(updatedProduct ->
            syncOptions.getCtpClient().execute(updateCommandFunction.apply(updatedProduct, batch)));
    }

    return completionStage;
}
```

File: src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: my mistake: not edge cases, but edge data. 
The fix still doesn't cover a case, where only the last update action is executed: if 501 of 502 actions are skipped - the test still success.
My point is: if we supply 502 action, we should ensure all 502 are targeted, or, at least, assert edge data: first, middle, last. In our case we also should assert all the data around limit edge, e.g. actions #498, #499, #500, #501

[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: What do you mean skipped? If you mean failed? then the version won't be updated..
So it's not possible that 501 actions have been skipped and only the last one is executed.

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java (hunk 3 above)
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/services/impl/BaseService.java]: 604de5d

File: src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: no, i mean they are not sent to the platform (filtered, forgot, malformed batching implementation and so on)

[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: BTW: i forgot to mention, that `batchElements()` tests also don't have corned cases testing (e.g edge values, close to batch  bounds, are included correctly)

[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: > no, i mean they are not sent to the platform (filtered, forgot, malformed batching implementation and so on)

Then it probably should be covered by a unit test.

I added more unit tests for batchElements covering uniform and non uniform batch splitting: 
b53af31

[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: more assertion: 
bf339f7

[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: still i can't understand why you don't want to create a full integration test, where you verify that all actions were delivered.... 
IMHO _shouldNotFail_ is not robust enough test.

[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: at least verify `product.version` is increased to 502- it is easy and cheap, so some 502 actions were delivered.

[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: > at least verify `product.version` is increased to 502- it is easy and cheap, so some 502 actions were delivered.

Maybe you haven't seen it, but I have done exactly that:
https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-4fcf28164f083f7b0c50764b1a938399R620

[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: @andrii-kovalenko-ct 
Added here a [`updateProduct_WithMoreThan500ImageAdditions_ShouldHaveAllNewImages `](https://github.com/commercetools/commercetools-sync-java/pull/202/files#diff-4fcf28164f083f7b0c50764b1a938399R619) IT which tests 600 addExternalImage update actions and checks if all are there ;)

[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java]: now i'm happy
